dissect/extfs/extfs.py

```diff
@@ -317,38 +317,38 @@ def crtime_ns(self):
         return _parse_ns_ts(time, time_extra)
 
     def listdir(self):
-        if self.filetype != stat.S_IFDIR:
-            raise NotADirectoryError(f"{self!r} is not a directory")
-
         if not self._dirlist:
-            dirs = {}
+            self._dirlist = {node.filename: node for node in self.iterdir()}
+        return self._dirlist
 
-            buf = io.BytesIO(self.open().read())
-            size_count = 0
+    dirlist = listdir
 
-            while size_count < self.size - 12:
-                direntry = self.extfs._dirtype(buf)
+    def iterdir(self):
+        if self.filetype != stat.S_IFDIR:
+            raise NotADirectoryError(f"{self!r} is not a directory")
 
-                # Sanity check if the direntry is valid
-                if direntry.inode < self.extfs.sb.s_inodes_count and direntry.inode > 1:
-                    fname = buf.read(direntry.name_len)
-                    fname = fname.decode(errors="surrogateescape")
-                    ftype = direntry.file_type if self.extfs._dirtype == c_ext.ext2_dir_entry_2 else None
+        buf = self.open()
+        offset = 0
 
-                    if ftype:
-                        ftype = FILETYPES[ftype]
+        while offset < self.size - 12:
+            direntry = self.extfs._dirtype(buf)
 
-                    inode = self.extfs.get_inode(direntry.inode, fname, ftype, parent=self, lazy=True)
-                    dirs[fname] = inode
+            if direntry.rec_len == 0:
```

Comment:
I think it would be good if we have a unit test that demonstrates how this change prevents an infinite loop? This also makes sure this issue will never return again...